Icinga2 fails to build

We could observe this different gcc versions (4.8.5 for sure) and not with others, so this is a bug in gcc which we will have to work around

```
cd /home/jenkins/workspace/icinga2-snapshot/rpm-centos-7-1binary/arch/x86_64/rpmbuild/BUILD/icinga2-2.8.0.139.g4fa42a8/lib/base && /usr/bin/cmake -E cmake_link_script CMakeFiles/base.dir/link.txt --verbose=1
/usr/lib64/ccache/c++  -fPIC -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -g -pthread -flto -fvisibility=hidden -std=c++0x -O2 -g -DNDEBUG   -Wl,--gc-sections -Wl,--no-export-dynamic -Bsymbolic-functions -Wl,--dynamic-list-cpp-typeinfo -Wl,--dynamic-list-data -flto -shared -Wl,-soname,libbase.so.2.8.0 -o ../../Bin/RelWithDebInfo/libbase.so.2.8.0 CMakeFiles/base.dir/application-version.cpp.o CMakeFiles/base.dir/base_unity.cpp.o  -L/home/jenkins/workspace/icinga2-snapshot/rpm-centos-7-1binary/arch/x86_64/rpmbuild/BUILD/icinga2-2.8.0.139.g4fa42a8/third-party/yajl -ldl -lboost_thread-mt -lboost_system-mt -lboost_program_options-mt -lboost_regex-mt -lssl -lcrypto ../../Bin/RelWithDebInfo/libyajl.so.2.1.0 ../../Bin/RelWithDebInfo/libmmatch.so.2.8.0 ../../Bin/RelWithDebInfo/libsocketpair.so.2.8.0 ../../Bin/RelWithDebInfo/libexecvpe.so.2.8.0 -Wl,-rpath,/home/jenkins/workspace/icinga2-snapshot/rpm-centos-7-1binary/arch/x86_64/rpmbuild/BUILD/icinga2-2.8.0.139.g4fa42a8/third-party/yajl:/home/jenkins/workspace/icinga2-snapshot/rpm-centos-7-1binary/arch/x86_64/rpmbuild/BUILD/icinga2-2.8.0.139.g4fa42a8/Bin/RelWithDebInfo: 
collect2: error: ld terminated with signal 11 [Segmentation fault], core dumped
make[2]: *** [Bin/RelWithDebInfo/libbase.so.2.8.0] Error 1
make[2]: Leaving directory `/home/jenkins/workspace/icinga2-snapshot/rpm-centos-7-1binary/arch/x86_64/rpmbuild/BUILD/icinga2-2.8.0.139.g4fa42a8'
make[1]: *** [lib/base/CMakeFiles/base.dir/all] Error 2
make[1]: Leaving directory `/home/jenkins/workspace/icinga2-snapshot/rpm-centos-7-1binary/arch/x86_64/rpmbuild/BUILD/icinga2-2.8.0.139.g4fa42a8'
make: *** [all] Error 2
error: Bad exit status from /var/tmp/rpm-tmp.ytmfTM (%build)
```

To blame is commit d5961bb76c65171e8eb868edf03c5330fa479a67

--- a/lib/base/functionwrapper.hpp
+++ b/lib/base/functionwrapper.hpp
@@ -23,6 +23,9 @@
 #include "base/i2-base.hpp"
 #include "base/value.hpp"
 #include <vector>
+#include <boost/function_types/function_type.hpp>
+#include <boost/function_types/parameter_types.hpp>
+#include <boost/function_types/result_type.hpp>
 #include <boost/function_types/function_arity.hpp>
 #include <type_traits>
 
@@ -31,7 +34,14 @@ using namespace std::placeholders;
 namespace icinga
 {
 
-inline std::function<Value (const std::vector<Value>&)> WrapFunction(const std::function<Value(const std::vector<Value>&)>& function)
+template<typename FuncType>
+typename std::enable_if<
+    std::is_class<FuncType>::value &&
+    std::is_same<typename boost::function_types::result_type<decltype(&FuncType::operator())>::type, Value>::value &&
+    std::is_same<typename boost::mpl::at_c<boost::function_types::parameter_types<decltype(&FuncType::operator())>, 1>::type, const std::vector<Value>&>::value &&
+    boost::function_types::function_arity<decltype(&FuncType::operator())>::value == 2,
+    std::function<Value (const std::vector<Value>&)>>::type
+WrapFunction(FuncType function)
 {
 	return function;
 }
@@ -43,12 +53,11 @@ inline std::function<Value (const std::vector<Value>&)> WrapFunction(void (*func
 		return Empty;
 	};
 }
+
 template<typename Return>
 std::function<Value (const std::vector<Value>&)> WrapFunction(Return (*function)(const std::vector<Value>&))
 {
-	return [function](const std::vector<Value>& arguments) {
-		return static_cast<Value>(function(arguments));
-	};
+	return std::bind(function, _1);
 }
 
 template <std::size_t... Indices>
@@ -69,52 +78,51 @@ struct build_indices<0> {
 template <std::size_t N>
 using BuildIndices = typename build_indices<N>::type;
 
-struct unpack_caller
+struct UnpackCaller
 {
 private:
 	template <typename FuncType, size_t... I>
-	auto call(FuncType f, const std::vector<Value>& args, indices<I...>) -> decltype(f(args[I]...))
+	auto Invoke(FuncType f, const std::vector<Value>& args, indices<I...>) -> decltype(f(args[I]...))
 	{
 		return f(args[I]...);
 	}
 
 public:
-	template <typename FuncType>
-	auto operator () (FuncType f, const std::vector<Value>& args)
-	    -> decltype(call(f, args, BuildIndices<boost::function_types::function_arity<typename boost::remove_pointer<FuncType>::type>::value>{}))
+	template <typename FuncType, int Arity>
+	auto operator() (FuncType f, const std::vector<Value>& args)
+	    -> decltype(Invoke(f, args, BuildIndices<Arity>{}))
 	{
-		return call(f, args, BuildIndices<boost::function_types::function_arity<typename boost::remove_pointer<FuncType>::type>::value>{});
+		return Invoke(f, args, BuildIndices<Arity>{});
 	}
 };
 
-enum class enabler_t {};
-
-template<bool T>
-using EnableIf = typename std::enable_if<T, enabler_t>::type;
-
-template<typename FuncType>
-std::function<Value (const std::vector<Value>&)> WrapFunction(FuncType function,
-    EnableIf<std::is_same<decltype(unpack_caller()(FuncType(), std::vector<Value>())), void>::value>* = 0)
+template<typename FuncType, int Arity, typename ReturnType>
+struct FunctionWrapper
 {
-	return [function](const std::vector<Value>& arguments) {
-		constexpr int arity = boost::function_types::function_arity<typename boost::remove_pointer<FuncType>::type>::value;
+	static Value Invoke(FuncType function, const std::vector<Value>& arguments)
+	{
+		return UnpackCaller().operator()<FuncType, Arity>(function, arguments);
+	}
+};
 
-		if (arguments.size() < arity)
-			BOOST_THROW_EXCEPTION(std::invalid_argument("Too few arguments for function."));
-		else if (arguments.size() > arity)
-			BOOST_THROW_EXCEPTION(std::invalid_argument("Too many arguments for function."));
-
-		unpack_caller()(function, arguments);
+template<typename FuncType, int Arity>
+struct FunctionWrapper<FuncType, Arity, void>
+{
+	static Value Invoke(FuncType function, const std::vector<Value>& arguments)
+	{
+		UnpackCaller().operator()<FuncType, Arity>(function, arguments);
 		return Empty;
-	};
-}
+	}
+};
 
 template<typename FuncType>
-std::function<Value (const std::vector<Value>&)> WrapFunction(FuncType function,
-		EnableIf<!std::is_same<decltype(unpack_caller()(FuncType(), std::vector<Value>())), void>::value>* = 0)
+typename std::enable_if<
+    std::is_function<typename std::remove_pointer<FuncType>::type>::value && !std::is_same<FuncType, Value(*)(const std::vector<Value>&)>::value,
+    std::function<Value (const std::vector<Value>&)>>::type
+WrapFunction(FuncType function)
 {
 	return [function](const std::vector<Value>& arguments) {
-		constexpr int arity = boost::function_types::function_arity<typename boost::remove_pointer<FuncType>::type>::value;
+		constexpr size_t arity = boost::function_types::function_arity<typename std::remove_pointer<FuncType>::type>::value;
 
 		if (arity > 0) {
 			if (arguments.size() < arity)
@@ -123,7 +131,36 @@ std::function<Value (const std::vector<Value>&)> WrapFunction(FuncType function,
 				BOOST_THROW_EXCEPTION(std::invalid_argument("Too many arguments for function."));
 		}
 
-		return unpack_caller()(function, arguments);
+		using ReturnType = decltype(UnpackCaller().operator()<FuncType, arity>(*static_cast<FuncType *>(nullptr), std::vector<Value>()));
+
+		return FunctionWrapper<FuncType, arity, ReturnType>::Invoke(function, arguments);
+	};
+}
+
+template<typename FuncType>
+typename std::enable_if<
+    std::is_class<FuncType>::value &&
+    !(std::is_same<typename boost::function_types::result_type<decltype(&FuncType::operator())>::type, Value>::value &&
+    std::is_same<typename boost::mpl::at_c<boost::function_types::parameter_types<decltype(&FuncType::operator())>, 1>::type, const std::vector<Value>&>::value &&
+    boost::function_types::function_arity<decltype(&FuncType::operator())>::value == 2),
+    std::function<Value (const std::vector<Value>&)>>::type
+WrapFunction(FuncType function)
+{
+	using FuncTypeInvoker = decltype(&FuncType::operator());
+
+	return [function](const std::vector<Value>& arguments) {
+		constexpr size_t arity = boost::function_types::function_arity<FuncTypeInvoker>::value - 1;
+
+		if (arity > 0) {
+			if (arguments.size() < arity)
+				BOOST_THROW_EXCEPTION(std::invalid_argument("Too few arguments for function."));
+			else if (arguments.size() > arity)
+				BOOST_THROW_EXCEPTION(std::invalid_argument("Too many arguments for function."));
+		}
+
+		using ReturnType = decltype(UnpackCaller().operator()<FuncType, arity>(*static_cast<FuncType *>(nullptr), std::vector<Value>()));
+
+		return FunctionWrapper<FuncType, arity, ReturnType>::Invoke(function, arguments);
 	};
 }
 

--- a/lib/config/vmops.hpp
+++ b/lib/config/vmops.hpp
@@ -113,7 +113,7 @@ public:
 	{
 		auto evaluatedClosedVars = EvaluateClosedVars(frame, closedVars);
 
-		auto wrapper = [argNames, evaluatedClosedVars, expression](const std::vector<Value>& arguments) {
+		auto wrapper = [argNames, evaluatedClosedVars, expression](const std::vector<Value>& arguments) -> Value {
 			if (arguments.size() < argNames.size())
 				BOOST_THROW_EXCEPTION(std::invalid_argument("Too few arguments for function"));
 

--- a/lib/icinga/comment.cpp
+++ b/lib/icinga/comment.cpp
@@ -99,7 +99,7 @@ void Comment::Start(bool runtimeCreated)
 
 	static boost::once_flag once = BOOST_ONCE_INIT;
 
-	boost::call_once(once, []() {
+	boost::call_once(once, [this]() {
 		l_CommentsExpireTimer = new Timer();
 		l_CommentsExpireTimer->SetInterval(60);
 		l_CommentsExpireTimer->OnTimerExpired.connect(std::bind(&Comment::CommentsExpireTimerHandler));

--- a/lib/icinga/downtime.cpp
+++ b/lib/icinga/downtime.cpp
@@ -103,7 +103,7 @@ void Downtime::Start(bool runtimeCreated)
 
 	static boost::once_flag once = BOOST_ONCE_INIT;
 
-	boost::call_once(once, []() {
+	boost::call_once(once, [this]() {
 		l_DowntimesStartTimer = new Timer();
 		l_DowntimesStartTimer->SetInterval(5);
 		l_DowntimesStartTimer->OnTimerExpired.connect(std::bind(&Downtime::DowntimesStartTimerHandler));

--- a/lib/icinga/macroprocessor.cpp
+++ b/lib/icinga/macroprocessor.cpp
@@ -180,30 +180,6 @@ bool MacroProcessor::ResolveMacro(const String& macro, const ResolverList& resol
 	return false;
 }
 
-Value MacroProcessor::InternalResolveMacrosShim(const std::vector<Value>& args, const ResolverList& resolvers,
-    const CheckResult::Ptr& cr, const MacroProcessor::EscapeCallback& escapeFn, const Dictionary::Ptr& resolvedMacros,
-    bool useResolvedMacros, int recursionLevel)
-{
-	if (args.size() < 1)
-		BOOST_THROW_EXCEPTION(std::invalid_argument("Too few arguments for function"));
-
-	String missingMacro;
-
-	return MacroProcessor::InternalResolveMacros(args[0], resolvers, cr, &missingMacro, escapeFn,
-	    resolvedMacros, useResolvedMacros, recursionLevel);
-}
-
-Value MacroProcessor::InternalResolveArgumentsShim(const std::vector<Value>& args, const ResolverList& resolvers,
-    const CheckResult::Ptr& cr, const Dictionary::Ptr& resolvedMacros,
-    bool useResolvedMacros, int recursionLevel)
-{
-	if (args.size() < 2)
-		BOOST_THROW_EXCEPTION(std::invalid_argument("Too few arguments for function"));
-
-	return MacroProcessor::ResolveArguments(args[0], args[1], resolvers, cr,
-	    resolvedMacros, useResolvedMacros, recursionLevel);
-}
-
 Value MacroProcessor::EvaluateFunction(const Function::Ptr& func, const ResolverList& resolvers,
     const CheckResult::Ptr& cr, const MacroProcessor::EscapeCallback& escapeFn,
     const Dictionary::Ptr& resolvedMacros, bool useResolvedMacros, int recursionLevel)
@@ -214,12 +190,27 @@ Value MacroProcessor::EvaluateFunction(const Function::Ptr& func, const Resolver
 		resolvers_this->Set(resolver.first, resolver.second);
 	}
 
-	resolvers_this->Set("macro", new Function("macro (temporary)", std::bind(&MacroProcessor::InternalResolveMacrosShim,
-	    _1, std::cref(resolvers), cr, MacroProcessor::EscapeCallback(), resolvedMacros, useResolvedMacros,
-	    recursionLevel + 1), { "str" }));
-	resolvers_this->Set("resolve_arguments", new Function("resolve_arguments (temporary)", std::bind(&MacroProcessor::InternalResolveArgumentsShim,
-	    _1, std::cref(resolvers), cr, resolvedMacros, useResolvedMacros,
-	    recursionLevel + 1)));
+	auto internalResolveMacrosShim = [resolvers, cr, resolvedMacros, useResolvedMacros, recursionLevel](const std::vector<Value>& args) {
+		if (args.size() < 1)
+			BOOST_THROW_EXCEPTION(std::invalid_argument("Too few arguments for function"));
+
+		String missingMacro;
+
+		return MacroProcessor::InternalResolveMacros(args[0], resolvers, cr, &missingMacro, MacroProcessor::EscapeCallback(),
+		    resolvedMacros, useResolvedMacros, recursionLevel);
+	};
+
+	resolvers_this->Set("macro", new Function("macro (temporary)", internalResolveMacrosShim, { "str" }));
+
+	auto internalResolveArgumentsShim = [resolvers, cr, resolvedMacros, useResolvedMacros, recursionLevel](const std::vector<Value>& args) {
+		if (args.size() < 2)
+			BOOST_THROW_EXCEPTION(std::invalid_argument("Too few arguments for function"));
+
+		return MacroProcessor::ResolveArguments(args[0], args[1], resolvers, cr,
+		    resolvedMacros, useResolvedMacros, recursionLevel + 1);
+	};
+
+	resolvers_this->Set("resolve_arguments", new Function("resolve_arguments (temporary)", internalResolveArgumentsShim, { "command", "args" }));
 
 	return func->InvokeThis(resolvers_this);
 }

--- a/lib/icinga/macroprocessor.hpp
+++ b/lib/icinga/macroprocessor.hpp
@@ -63,12 +63,6 @@ private:
 	    String *missingMacro, const EscapeCallback& escapeFn,
 	    const Dictionary::Ptr& resolvedMacros, bool useResolvedMacros,
 	    int recursionLevel = 0);
-	static Value InternalResolveMacrosShim(const std::vector<Value>& args, const ResolverList& resolvers,
-	    const CheckResult::Ptr& cr, const MacroProcessor::EscapeCallback& escapeFn,
-            const Dictionary::Ptr& resolvedMacros, bool useResolvedMacros, int recursionLevel);
-	static Value InternalResolveArgumentsShim(const std::vector<Value>& args, const ResolverList& resolvers,
-	    const CheckResult::Ptr& cr, const Dictionary::Ptr& resolvedMacros,
-	    bool useResolvedMacros, int recursionLevel);
 	static Value EvaluateFunction(const Function::Ptr& func, const ResolverList& resolvers,
 	    const CheckResult::Ptr& cr, const MacroProcessor::EscapeCallback& escapeFn,
 	    const Dictionary::Ptr& resolvedMacros, bool useResolvedMacros, int recursionLevel);

--- a/lib/icinga/scheduleddowntime.cpp
+++ b/lib/icinga/scheduleddowntime.cpp
@@ -91,7 +91,7 @@ void ScheduledDowntime::Start(bool runtimeCreated)
 
 	static boost::once_flag once = BOOST_ONCE_INIT;
 
-	boost::call_once(once, []() {
+	boost::call_once(once, [this]() {
 		l_Timer = new Timer();
 		l_Timer->SetInterval(60);
 		l_Timer->OnTimerExpired.connect(std::bind(&ScheduledDowntime::TimerProc));

--- a/lib/icinga/timeperiod.cpp
+++ b/lib/icinga/timeperiod.cpp
@@ -40,7 +40,7 @@ void TimePeriod::Start(bool runtimeCreated)
 
 	static boost::once_flag once = BOOST_ONCE_INIT;
 
-	boost::call_once(once, []() {
+	boost::call_once(once, [this]() {
 		l_UpdateTimer = new Timer();
 		l_UpdateTimer->SetInterval(300);
 		l_UpdateTimer->OnTimerExpired.connect(std::bind(&TimePeriod::UpdateTimerHandler));

--- a/test/remote-url.cpp
+++ b/test/remote-url.cpp
@@ -57,8 +57,8 @@ BOOST_AUTO_TEST_CASE(get_and_set)
 	std::vector<String> v1 { "hip", "hip", "hurra" };
 	std::vector<String> v2 { "äü^ä+#ül-" };
 	std::vector<String> v3 { "1", "2" };
-	m.emplace("shout", v1);
-	m.emplace("sonderzeichen", v2);
+	m.insert(std::make_pair("shout", v1));
+	m.insert(std::make_pair("sonderzeichen", v2));
 	url->SetQuery(m);
 	url->SetQueryElements("count", v3);
 	url->AddQueryElement("count", "3");

